### PR TITLE
tech-story: [M3-8940] - Dev Tools fixes and improvements

### PR DESCRIPTION
**Pull request by abailly-akamai**
## Description 📝
Long overdue updates to our beloved Dev Tools. This PR does not touch any client code. Brings a lot of improvements and fixes to presets and mocking.

> [!NOTE]
> Code quality/readability/reusability is not as high as it would be for client code, tho I tried best to achieve that while working quickly. The goal aims to help the developer experience first 💨 

## Changes  🔄
- Allow JSON/nested feature flags toggling (recursively)
- Fix DevTools panel overflow
- Improve re-haul Account/Profile mocking
- Style improvements

## Target release date 🗓️

Please specify a release date (and environment, if applicable) to guarantee timely review of this PR. If exact date is not known, please approximate and update it as needed.

## Preview 📷
Screenshot 2024-11-26 at 14 38 03

Screenshot 2024-11-26 at 14 38 18

### Verification steps
- Play with MSW, verify 
  - Nested feature flags 
  - Profile and accounts
  - No regressions in saving, storing mock data etc

## As an Author, I have considered 🤔

- 👀 Doing a self review
- ❔ Our [contribution guidelines](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md)
- 🤏 Splitting feature into small PRs
- ➕ Adding a [changeset](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md#writing-a-changeset)
- 🧪 Providing/improving test coverage
- 🔐 Removing all sensitive information from the code and PR description
- 🚩 Using a feature flag to protect the release
- 👣 Providing comprehensive reproduction steps
- 📑 Providing or updating our documentation
- 🕛 Scheduling a pair reviewing session
- 📱 Providing mobile support
- ♿  Providing accessibility support

- [x] I have read and considered all applicable items listed above.

## As an Author, before moving this PR from Draft to Open, I confirmed ✅

- [x] All unit tests are passing
- [x] TypeScript compilation succeeded without errors
- [x] Code passes all linting rules

### CODE REVIEW

File: packages/manager/src/dev-tools/components/JsonTextArea.tsx [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+  height,
+  label,
+  name,
+  onChange,
+  value,
+}: JsonTextAreaProps) => {
+  const [rawText, setRawText] = React.useState(JSON.stringify(value, null, 2));
+
+  const debouncedUpdate = React.useCallback(
+    React.useMemo(
```
Comment (jaalah-akamai) [inline packages/manager/src/dev-tools/components/JsonTextArea.tsx:25]: I'm thinking this double memoization is a mistake?

**Review by jaalah-akamai — APPROVED** [review]
Nice work overall, I actually used the account preset for my testing today for OBJ. 🔥

File: packages/manager/src/dev-tools/FeatureFlagTool.tsx [hunk 2]
```diff
@@ -81,15 +139,27 @@ export const FeatureFlagTool = withFeatureFlagProvider(() => {
...
+      // Only update the specific property that changed
+      updatedFlags[parentKey] = {
+        ...currentParentValue, // Keep original LD values
+        ...existingValues, // Apply any existing stored overrides
+        [childKey]: updatedValue, // Apply the new change
+      };
+    }
+
     dispatch(setMockFeatureFlags(updatedFlags));
     setStorage(MOCK_FEATURE_FLAGS_STORAGE_KEY, JSON.stringify(updatedFlags));
```
[the remark this answers is not in the document]
Comment (abailly-akamai) [reply packages/manager/src/dev-tools/FeatureFlagTool.tsx:171]: absolutely, done

File: packages/manager/src/dev-tools/FeatureFlagTool.tsx [hunk 3]
```diff
@@ -81,15 +139,27 @@ export const FeatureFlagTool = withFeatureFlagProvider(() => {
...
-    };
+    const updatedValue = e.target.checked;
+    const storedFlags = getStorage(MOCK_FEATURE_FLAGS_STORAGE_KEY) || {};
+
+    const flagParts = flag.split('.');
+    const updatedFlags = { ...storedFlags };
+
+    if (flagParts.length === 1) {
+      updatedFlags[flag] = updatedValue;
+    } else {
```
[the remark this answers is not in the document]
Comment (abailly-akamai) [reply packages/manager/src/dev-tools/FeatureFlagTool.tsx:152]: gonna keep this as is, it's readable and neither is a "bad case", just nested VS non-nested flags

File: packages/manager/src/dev-tools/ServiceWorkerTool.tsx [hunk 4]
```diff
@@ -72,6 +83,22 @@ export const ServiceWorkerTool = () => {
...
 
+  React.useEffect(() => {
+    const currentAccountData = getCustomAccountData();
+    const currentProfileData = getCustomProfileData();
+    const hasCustomAccountChanges =
+      JSON.stringify(currentAccountData) !== JSON.stringify(customAccountData);
+    const hasCustomProfileChanges =
+      JSON.stringify(currentProfileData) !== JSON.stringify(customProfileData);
+
+    if (hasCustomAccountChanges || hasCustomProfileChanges) {
```
[the remark this answers is not in the document]
Comment (abailly-akamai) [reply packages/manager/src/dev-tools/ServiceWorkerTool.tsx:94]: yeah... i was looking for a util, realized we don't have one, and opted against importing from CP cause.. who knows what's gonna happen to that 😨 , and i want to keep dev tools separated enough from

File: packages/manager/src/dev-tools/components/ExtraPresetAccount.tsx [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+    onTogglePreset(e, 'account:custom');
+  };
+
+  React.useEffect(() => {
+    if (!isEnabled) {
+      setCustomAccountData(null);
+    } else if (isEnabled && customAccountData) {
+      setCustomAccountData(customAccountData);
+    }
+  }, [isEnabled, customAccountData]);
```
[the remark this answers is not in the document]
Comment (abailly-akamai) [reply packages/manager/src/dev-tools/components/ExtraPresetAccount.tsx:99]: yup! good catch, done

File: packages/manager/src/dev-tools/components/ExtraPresetProfile.tsx [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+      setFormData((prev) => ({
+        ...prev,
+        ...customProfileData,
+      }));
+    } else {
+      setFormData({
+        ...profileFactory.build(),
+      });
+    }
+  }, [customProfileData, isEnabled]);
```
[the remark this answers is not in the document]
Comment (abailly-akamai) [reply packages/manager/src/dev-tools/components/ExtraPresetProfile.tsx]: done!

File: packages/manager/src/dev-tools/components/ExtraPresetProfile.tsx [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+
+    const newFormData = {
+      ...formData,
+      [name]: value,
+    };
+    setFormData(newFormData);
+
+    if (isEnabled) {
+      onFormChange?.(newFormData);
+    }
```
[the remark this answers is not in the document]
Comment (abailly-akamai) [reply packages/manager/src/dev-tools/components/ExtraPresetProfile.tsx]: much cleaner, thx!

File: packages/manager/src/dev-tools/components/JsonTextArea.tsx (hunk 1 above)
Comment (abailly-akamai) [reply packages/manager/src/dev-tools/components/JsonTextArea.tsx]: indeed! fixed

**Review by jdamore-linode — APPROVED** [review]
This is extremely cool @abailly-akamai, thanks for taking the time to add all this!

One technical question -- I was just messing around with the new feature flag work and tried to add a new JSON flag, but it's showing up in the dev tools as a checkbox. Can you spot what I did wrong?

```typescript
// featureFlags.ts
interface JustMessingAroundFlag extends BaseFeatureFlag {
  optionA: boolean;
  optionB: boolean;
}

export interface Flags {
  justMessingAround: JustMessingAroundFlag;
  // ...
}
```
```typescript
// FeatureFlagTool.tsx
const options: { flag: keyof Flags; label: string }[] = [
  { flag: 'justMessingAround', label: 'Just Messing Around' },
  // ...
];
```
Screenshot 2024-12-04 at 11 47 57 AM

Existing flags seem to work fine -- for example, I added the `cloudManagerDesignUpdatesBanner` flag to the array in `FeatureFlagTool.tsx`, and it showed up as expected, so I'm thinking I must be doing something wrong with the way I've set up my flag.

Either way, these are all such nice/powerful improvements! Thanks again Alban!

**Comment by abailly-akamai** [comment]
@jdamore-linode could it simply be the fact that `justMessingAround` is typed but does not return any actual value from LD?

**Comment by jdamore-linode** [comment]
> @jdamore-linode could it simply be the fact that `justMessingAround` is typed but does not return any actual value from LD?

Ah thanks! I didn't realize they had to be tied to the actual Launch Darkly data since it's being mocked.

**Pull request merged**